apps/aecore/test/aec_conductor_tests.erl

```diff
@@ -373,17 +379,147 @@ test_block_publishing() ->
     ?assertEqual([], flush_gproc()),
     ok.
 
+%%%===================================================================
+%%% Micro block signing tests
+%%%===================================================================
+
+generation_test_() ->
+    {foreach,
+     fun() ->
+             TmpKeysDir = setup_common(),
+             {ok, _} = ?TEST_MODULE:start_link([{autostart, false}]),
+             meck:new(aec_headers, [passthrough]),
+             meck:new(aec_blocks, [passthrough]),
+             meck:expect(aec_headers, validate_key_block_header, fun(_) -> ok end),
+             meck:expect(aec_headers, validate_micro_block_header, fun(_) -> ok end),
+             meck:expect(aec_blocks, validate_key_block, fun(_) -> ok end),
+             meck:expect(aec_blocks, validate_micro_block, fun(_) -> ok end),
+             TmpKeysDir
+     end,
+     fun(TmpKeysDir) ->
+             teardown_common(TmpKeysDir),
+             meck:unload(aec_headers),
+             meck:unload(aec_blocks),
+             ok
+     end,
+     [
+        {"Start signing after mined block", fun test_mined_block_signing/0},
+        {"Start signing after two mined block", fun test_two_mined_block_signing/0},
+        {"Start signing after received block", fun test_received_block_signing/0}
+     ]}.
+
+test_mined_block_signing() ->
+    Keys = beneficiary_keys(),
+    true = aec_events:subscribe(block_created),
+    true = aec_events:subscribe(micro_block_created),
+
+    ?TEST_MODULE:start_mining(),
+    assert_leader(false),
+    assert_generation_state(stopped),
+    KeyBlock = wait_for_block_created(),
+    assert_leader(true),
+    assert_generation_state(running),
+
+    ok = aec_tx_pool:push(tx(Keys)),
+
+    MicroBlock = wait_for_micro_block_created(),
+    {ok, KeyHash} = aec_blocks:hash_internal_representation(KeyBlock),
+    ?assertEqual(KeyHash, aec_blocks:prev_key_hash(MicroBlock)),
+    ok.
+
+test_two_mined_block_signing() ->
+    Keys = beneficiary_keys(),
+    true = aec_events:subscribe(block_created),
+    true = aec_events:subscribe(micro_block_created),
+
+    ?TEST_MODULE:start_mining(),
+    assert_leader(false),
+    assert_generation_state(stopped),
+    KeyBlock1 = wait_for_block_created(),
+    assert_leader(true),
+    assert_generation_state(running),
+    KeyBlock2 = wait_for_block_created(),
+    assert_leader(true),
+    assert_generation_state(running),
+
+    ok = aec_tx_pool:push(tx(Keys)),
+
+    MicroBlock = wait_for_micro_block_created(),
+    {ok, KeyHash1} = aec_blocks:hash_internal_representation(KeyBlock1),
+    {ok, KeyHash2} = aec_blocks:hash_internal_representation(KeyBlock2),
+    ?assertEqual(KeyHash1, aec_blocks:prev_key_hash(KeyBlock2)),
+    ?assertEqual(KeyHash2, aec_blocks:prev_key_hash(MicroBlock)),
+    ok.
+
+test_received_block_signing() ->
+    Keys = beneficiary_keys(),
+    meck:new(aec_mining, [passthrough]),
```

Comment:
If test fails after this line but before the corresponding unload, does any teardown remove this mock? if not, this would affect validity of results of following tests.

Reply:
🔬👍  I'll move it teardown